### PR TITLE
Allow WithTransform()'s function to accept a nil value

### PR DESCRIPTION
**Pull request by jim-slattery-rs**
Add a unit test to verify this use case.

### CODE REVIEW

**Comment by onsi** [comment]
awesome, thanks!

File: matchers/with_transform.go [hunk 1]
```diff
@@ -40,15 +40,25 @@ func NewWithTransformMatcher(transform interface{}, matcher types.GomegaMatcher)
...
+			// return error if actual's type is incompatible with Transform function's argument type
+			actualType := reflect.TypeOf(actual)
+			if !actualType.AssignableTo(m.transformArgType) {
+				return false, fmt.Errorf("Transform function expects '%s' but we have '%s'", m.transformArgType, actualType)
+			}
+			param = reflect.ValueOf(actual)
+		} else {
+			// make a nil value of the expected type
+			param = reflect.New(m.transformArgType).Elem()
+		}
```
Comment (cbandy) [inline matchers/with_transform.go:56]: This is no longer checking that the `actual = nil` is appropriate for the transform function. This incorrectly succeeds:

```go
gomega.WithTransform(
	func(int) bool { return true },
	gomega.BeTrue(),
).Match(nil)
```

I recommend the following:

```go
	var param reflect.Value
	if actual == nil && m.transformArgType.Kind() == reflect.Interface {
		// The original type of actual is unknown, so there's no way to make its
		// reflect.Value. Create a nil of the transform argument, which is known.
		param = reflect.Zero(m.transformArgType)

	} else if actual != nil && reflect.TypeOf(actual).AssignableTo(m.transformArgType) {
		// The original type of actual is compatible with the transform argument.
		param = reflect.ValueOf(actual)

	} else {
		return false, fmt.Errorf("Transform function expects '%s' but we have '%T'", m.transformArgType, actual)
	}
```

Comment (onsi) [reply matchers/with_transform.go:56]: thanks @cbandy - are you up for submitting this as a PR?

Comment (cbandy) [reply matchers/with_transform.go:56]: Yep, no problem. Submitted #423.

**Pull request merged**